gcc/rust/typecheck/rust-tyty.cc

```diff
@@ -27,13 +27,41 @@
 #include "rust-tyty-cast.h"
 #include "rust-hir-map.h"
 #include "rust-substitution-mapper.h"
+#include "rust-hir-trait-ref.h"
+#include "rust-hir-type-bounds.h"
 
 extern ::Backend *
 rust_get_backend ();
 
 namespace Rust {
 namespace TyTy {
 
+bool
+BaseType::satisfies_bound (const TypeBoundPredicate &predicate) const
+{
+  const Resolver::TraitReference *query = predicate.get ();
+  for (auto &bound : specified_bounds)
+    {
+      const Resolver::TraitReference *item = bound.get ();
+      bool found = item->get_mappings ().get_defid ()
+		   == query->get_mappings ().get_defid ();
+      if (found)
+	return true;
+    }
+
+  std::vector<std::reference_wrapper<Resolver::TraitReference>> probed
+    = Resolver::TypeBoundsProbe::Probe (this);
+  for (const Resolver::TraitReference &bound : probed)
+    {
+      bool found = bound.get_mappings ().get_defid ()
+		   == query->get_mappings ().get_defid ();
+      if (found)
+	return true;
+    }
```

Comment:
I'm not sure I'm understanding this code correctly. Does this mean that as soon as you find a matching bound, the function returns true? In that case, is it intended (for now) that multiple bounds are not verified correctly?

For example, the following example compiles succesfully:

```rust
trait Foo {
    fn default() -> i32;
}

trait Bar {
    fn not_default() -> i32;
}

struct Test(i32);

impl Foo for Test {
    fn default() -> i32 {
        1234
    }
}

fn type_bound_test<T: Foo + Bar>() -> i32 { // <T: Bar + Foo> also compiles
    T::default()
}

fn main() {
    let a = type_bound_test::<Test>();
}
```

Reply:
Let me add that test case to verify this.

But the code is split up into 2 parts this function is about checking each bound the driver of this is actually here: 

https://github.com/Rust-GCC/gccrs/pull/611/files#diff-03fbb543dd5e74fc2332fa2e5c9c9a91ed6d46d86e7bf0310a8796fae439d26cR256-R275

So we iterate the bounds of the TypePaameter then verify that each of these are satisfied by the substitution.

Reply:
Ah, right thanks. I see how the code works exactly now. Glad that multiple bounds are working :D Good job!